client-go: changes config.Timeout field semantic

p0lyn0mial opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?
/kind bug
<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature
/kind design

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:
This PR uses context for cancellation instead of setting a timeout on the HTTP client.

It turns out that setting a timeout on HTTP client affect watch requests.
For example, with a 10 second timeout watch requests are being re-established exactly after 10 seconds even though the default request timeout for them is ~5 minutes (informers).

This is because if multiple timeouts were set, the stdlib picks the smaller timeout to be applied, leaving other useless.
For more details see https://github.com/golang/go/blob/a937729c2c2f6950a32bc5cd0f5b88700882f078/src/net/http/client.go#L364

This PR preserves the previous behavior for all requests except the watch requests. With this PR a timeout set by `ListOptions` on watches is respected - used heavily by the informers. In particular when a global timeout via `config.Timeout` was set and no other timeout was specified then watch requests will be terminated after `config.Timeout`, just like before.

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note

```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

p0lyn0mial committed:
client-go: rest client changes config.Timeout field semantic

It turns out that setting a timeout on HTTP client affect watch requests.
For example, with a 10 second timeout watch requests are being re-established exactly after 10 seconds even though the default request timeout for them is ~5 minutes.

This is because if multiple timeouts were set, the stdlib picks the smaller timeout to be applied, leaving other useless.
For more details see https://github.com/golang/go/blob/a937729c2c2f6950a32bc5cd0f5b88700882f078/src/net/http/client.go#L364

Instead of setting a timeout on the HTTP client we should use context for cancellation.

p0lyn0mial commented:
/assing @sttts @deads2k

sttts commented on staging/src/k8s.io/client-go/rest/client.go:
```diff
@@ -100,11 +100,14 @@ type RESTClient struct {
 // NewRESTClient creates a new RESTClient. This client performs generic REST functions
 // such as Get, Put, Post, and Delete on specified paths.
-func NewRESTClient(baseURL *url.URL, versionedAPIPath string, config ClientContentConfig, rateLimiter flowcontrol.RateLimiter, client *http.Client) (*RESTClient, error) {
+func NewRESTClient(baseURL *url.URL, versionedAPIPath string, config ClientContentConfig, rateLimiter flowcontrol.RateLimiter, client *http.Client, requestTimeout time.Duration) (*RESTClient, error) {
```
worth to avoid a breaking change through `NewRESTClientWithTimeout` ?

p0lyn0mial replied:
How many clients of this method do we have? 

I didn't have to fix many references, just two.

caesarxuchao replied:
Since client-go is a library, this change affects non k/k code as well.

sttts commented on staging/src/k8s.io/client-go/rest/client.go:
```diff
@@ -100,11 +100,14 @@ type RESTClient struct {
 	Client *http.Client
+
+	// RequestTimeout specifies a time limit for requests made by the HTTP Client
+	RequestTimeout time.Duration
```
does this have to be public?

p0lyn0mial replied:
no, I don't think so. Will change to private.

p0lyn0mial commented on test/integration/apimachinery/watch_restart_test.go:
```diff
@@ -275,3 +275,83 @@ func TestWatchRestartsIfTimeoutNotReached(t *testing.T) {
+
+	// validate
+	if len(res) > failedProbesTreshold {
+		t.Fatalf("%d%% of probes failed. That means some (all) watch requests ended sooner than we expected. The current treshold was set to %d%%"+
```
this message might look like this:

```
100% of probes failed. That means some (all) watch requests ended sooner than we expected. The current treshold was set to 33% We did 6 probes in total. The upper-bound time limit for a single watch reqeust was set to 11s. 6 probes ended before that time [10.000410361s 10.001010468s 10.000472694s 10.001261837s 10.002812845s 10.001133277s]
```

p0lyn0mial commented:
If this PR turns out to be viable then we don't need https://github.com/kubernetes/kubernetes/pull/100959

p0lyn0mial commented:
/retest

deads2k commented on staging/src/k8s.io/client-go/rest/config.go:
```diff
@@ -324,9 +324,12 @@ func RESTClientFor(config *Config) (*RESTClient, error) {
 	if transport != http.DefaultTransport {
 		httpClient = &http.Client{Transport: transport}
-		if config.Timeout > 0 {
-			httpClient.Timeout = config.Timeout
```
With this removed, how do I create a rest.Config that has a client-side, context based timeout on every request?  I was fairly sure this was exactly the right spot to set up such a timeout.

p0lyn0mial replied:
>  how do I create a rest.Config that has a client-side, context based timeout on every request?

In exactly the same way as before. This behavior hasn't changed.

deads2k commented:
A global timeout, per client, enforced client-side, seems very reasonable to me and matches what the existing code does.  The current problem we face appears to be using a client with an explicitly configured, client-enforced timeout for a request we don't want to have the timeout enforced on.  This appears to be an "us" problem, not a client problem.

Why wouldn't the delegating authenticator simply create the client with the configuration it wants?  Clearly it does not want watches terminating every 10 seconds.

p0lyn0mial commented:
> A global timeout, per client, enforced client-side, seems very reasonable to me and matches what the existing code does

This PR sets a global, per client timeout too. That behaviour hasn't changed.
What has changed is that it respects a timeout set by watch requests.


>The current problem we face appears to be using a client with an explicitly configured, client-enforced timeout for a request we don't want to have the timeout enforced on.

It seems that the current problem we face is that the timeout for watch request is not respected at all.

> Why wouldn't the delegating authenticator simply create the client with the configuration it wants? Clearly it does not want watches terminating every 10 seconds.

I think that it would be good if the delegating authenticator could set a short timeout for `SAR` requests and a longer timeout for watch requests. This PR allows thath.

liggitt commented on staging/src/k8s.io/client-go/rest/config.go:
```diff
@@ -324,9 +324,12 @@ func RESTClientFor(config *Config) (*RESTClient, error) {
-			httpClient.Timeout = config.Timeout
-		}
+		// do not set a timeout on the http client, instead use context for cancellation
+		// if multiple timeouts were set, the request will pick the smaller timeout to be applied, leaving other useless.
```
> if multiple timeouts were set, the request will pick the smaller timeout to be applied, leaving other useless

if multiple timeouts were set, I would *expect* the smallest to rule

p0lyn0mial replied:
if this is what we want for client-go then sure. As it is today sharing the same client for normal requests and informers and setting a `conig.Timeout` to `10 seconds` will terminate watches created by informers every `10 seconds`.

smarterclayton replied:
You should not have to create two clients to use an informer.  client-go should let you specifically set a timeout for long running requests distinct from normal requests, and it should be clear what it does.

p0lyn0mial committed:
update client gen

p0lyn0mial committed:
generated

p0lyn0mial committed:
fixes a failing integration cmd test

Before the "check --request-timeout on 'get pod' with --watch" test in request-timeout.sh expected to get "Unable to decode an event from the watch stream: net/http: request canceled (Client.Timeout or context cancellation while reading body)".
It was changed to expect "Unable to decode an event from the watch stream: context deadline exceeded" instead.

deads2k commented:
> A global timeout, per client, enforced client-side, seems very reasonable to me and matches what the existing code does. The current problem we face appears to be using a client with an explicitly configured, client-enforced timeout for a request we don't want to have the timeout enforced on. This appears to be an "us" problem, not a client problem.
--deads2

> > if multiple timeouts were set, the request will pick the smaller timeout to be applied, leaving other useless
> 
> if multiple timeouts were set, I would _expect_ the smallest to rule
--liggitt


@p0lyn0mial I think what may be happening is that you're viewing the timeout hierarchy differently.  I expect the shortest timeout to rule and I want that timeout honored client-side, not just as a query parameter.  I suspect you may be viewing it as a longer per-request timeout would override a shorter global timeout, but I don't agree that is expected or desireable.

I may be open to that idea *in addition* to the current mechanism, but not in lieu of it.  I think the current bug is on the part of the delegating authenticator.

fedebongio commented:
/assign @sttts @deads2k @caesarxuchao 
/triage accepted

p0lyn0mial commented:
spoke to @sttts, we are going to open a new PR and propose a new field for `client-go` that would allow for setting a short timeout for all requests except watch requets.

smarterclayton commented:
I like stefan's suggestion.

I expected `rest.Config` to have

```
// Timeout is optional and will terminate all requests after this time, even long running requests.
Timeout int
// RequestTimeout terminates requests after this amount of time. It applies to requests that are not long running like watches, streams, or exec sessions.  Setting RequestTimeout longer than Timeout will have no effect.
RequestTimeout int
```

enj commented:
/remove-sig auth

p0lyn0mial commented:
/remove-lifecycle stal

aojea commented:
/cc

dims commented:
Is this PR still needed, please rebase if so (or we can close it?)

caesarxuchao commented:
/unassign

liggitt commented:
closing due to inactivity
